### PR TITLE
PERF-#360: Implement asyncronous get request

### LINKED ISSUE #360: Improve performance for get request
Get data as soon as they are ready

### PR DESCRIPTION
**Pull request by Retribution98**
## What do these changes do?

- [x] first commit message and PR title follow format outlined [here](https://unidist.readthedocs.io/en/latest/developer/contributing.html#commit-message-formatting)
  > **_NOTE:_**  If you edit the PR title to match this format, you need to add another commit (even if it's empty) or amend your last commit for the CI job that checks the PR title to pick up the new PR title.
- [x] passes `flake8 .`
- [x] passes `black .`
- [x] signed commit with `git commit -s` 
- [x] Resolves #360 ? 
- [x] tests added and passing
- [x] module layout described at `docs/developer/architecture.rst` is up-to-date

### CODE REVIEW

File: unidist/core/backends/mpi/core/communication.py [hunk 1]
```diff
@@ -374,6 +374,18 @@ def mpi_recv_operation(comm):
     return op_type, status.Get_source()
 
 
+def mpi_iprobe_object(comm):
```
Comment (YarShev) [inline unidist/core/backends/mpi/core/communication.py:377]: docstring

File: unidist/core/backends/mpi/core/shared_object_store.py [hunk 2]
```diff
@@ -431,18 +431,17 @@ def _write_to_shared_buffer(self, data_id, reservation_data, serialized_data):
         for i, raw_buffer in enumerate(raw_buffers):
             raw_buffer_first_index = last_prev_index
             raw_buffer_len = len(raw_buffer)
-            raw_buffer_last_index = raw_buffer_first_index + len(raw_buffer)
-            if s_data_last_index > last_index:
+            last_prev_index = raw_buffer_first_index + len(raw_buffer)
```
Comment (YarShev) [inline unidist/core/backends/mpi/core/shared_object_store.py:434]: I guess this is not related to the issue this PR resolves. Discard the changes in this file.

Comment (Retribution98) [reply unidist/core/backends/mpi/core/shared_object_store.py:434]: Why not? This isn't a big change, but it helps catch reserved memory overflows.

Comment (YarShev) [reply unidist/core/backends/mpi/core/shared_object_store.py:434]: If it helps to catch reserved memory overflows, how is this related to the get improvement? If it is not related, it is certainly a separate issue and should be addressed in a separate PR. We want to keep changes isolated in order for us to easily sort them out in future.

Comment (Retribution98) [reply unidist/core/backends/mpi/core/shared_object_store.py:434]: It's just renaming a variable!
Every change included in the PR was made while working on the issue. Every task can be divided into its simplest subtask, but not every subtask needs to be a separated issue.

File: unidist/core/backends/mpi/core/controller/api.py [hunk 3]
```diff
@@ -408,32 +408,34 @@ def get(data_ids):
     object
         A Python object.
     """
+    is_list = isinstance(data_ids, list)
```
Comment (YarShev) [inline unidist/core/backends/mpi/core/controller/api.py:411]:
```suggestion
    is_list = isinstance(data_ids, list)
    if not is_list:
        data_ids = [data_ids]
```

File: unidist/core/backends/mpi/core/controller/api.py [hunk 4]
```diff
@@ -408,32 +408,34 @@ def get(data_ids):
...
-        if isinstance(value, Exception):
-            raise value
+    if is_list:
+        remote_data_ids = [
+            data_id for data_id in data_ids if not local_store.contains(data_id)
+        ]
+    elif local_store.contains(data_ids):
+        remote_data_ids = []
+    else:
+        remote_data_ids = [data_ids]
```
Comment (YarShev) [inline unidist/core/backends/mpi/core/controller/api.py:421]:
```suggestion
    remote_data_ids = [
        data_id for data_id in data_ids if not local_store.contains(data_id)
    ]
```

File: unidist/core/backends/mpi/core/controller/api.py [hunk 5]
```diff
@@ -408,32 +408,34 @@ def get(data_ids):
...
 
-        return value
+    if remote_data_ids:
```
Comment (YarShev) [inline unidist/core/backends/mpi/core/controller/api.py:423]:
```suggestion
    # Remote data gets available in the local store inside `request_worker_data`
    if remote_data_ids:
```

File: unidist/core/backends/mpi/core/controller/api.py [hunk 6]
```diff
@@ -408,32 +408,34 @@ def get(data_ids):
...
-    is_list = isinstance(data_ids, list)
-    if not is_list:
-        data_ids = [data_ids]
-
-    values = [get_impl(data_id) for data_id in data_ids]
+    return_value = None
+    if is_list:
+        return_value = [local_store.get(data_id) for data_id in data_ids]
+    else:
+        return_value = local_store.get(data_ids)
```
Comment (YarShev) [inline unidist/core/backends/mpi/core/controller/api.py:432]:
```suggestion
    values = [local_store.get(data_id) for data_id in data_ids]
```

File: unidist/core/backends/mpi/core/controller/api.py [hunk 7]
```diff
@@ -408,32 +408,34 @@ def get(data_ids):
...
 
     # Initiate reference count based cleaup
     # if all the tasks were completed
     garbage_collector.regular_cleanup()
 
-    return values if is_list else values[0]
+    return return_value
```
Comment (YarShev) [inline unidist/core/backends/mpi/core/controller/api.py:438]:
```suggestion
    return values if is_list else values[0]
```

File: unidist/core/backends/mpi/core/controller/common.py [hunk 8]
```diff
@@ -107,7 +108,7 @@ def release_rank(self, rank):
         )
 
 
-def pull_data(comm, owner_rank):
+def pull_data(comm, owner_rank=None):
```
Comment (YarShev) [inline unidist/core/backends/mpi/core/controller/common.py:111]:
```suggestion
def pull_data(comm, owner_rank):
```
We can pass in MPI.ANY_SOURCE. The docstring should be elaborated with respect to this value.

File: unidist/core/backends/mpi/core/controller/common.py [hunk 9]
```diff
@@ -159,7 +166,7 @@ def pull_data(comm, owner_rank):
         raise ValueError("Unexpected package of data info!")
 
 
-def request_worker_data(data_id):
+def request_worker_data(data_ids):
```
Comment (YarShev) [inline unidist/core/backends/mpi/core/controller/common.py:169]: Fix the docstring.

File: unidist/core/backends/mpi/core/controller/common.py [hunk 10]
```diff
@@ -175,39 +182,41 @@ def request_worker_data(data_id):
...
+        operation_data = {
+            "source": mpi_state.global_rank,
+            "id": data_id,
+            # set `is_blocking_op` to `True` to tell a worker
+            # to send the data directly to the requester
+            # without any delay
+            "is_blocking_op": True,
+        }
+        # We use a blocking send here because we have to wait for
+        # completion of the communication, which is necessary for the pipeline to continue.
```
Comment (YarShev) [inline unidist/core/backends/mpi/core/controller/common.py:203]:
```suggestion
```

File: unidist/core/backends/mpi/core/controller/common.py [hunk 11]
```diff
@@ -175,39 +182,41 @@ def request_worker_data(data_id):
...
+        h_list = communication.isend_simple_operation(
+            mpi_state.comm,
+            operation_type,
+            operation_data,
+            owner_rank,
+        )
+        async_operations.extend(h_list)
```
Comment (YarShev) [inline unidist/core/backends/mpi/core/controller/common.py:210]: We should probably wait on these async requests.

File: unidist/core/backends/mpi/core/controller/common.py [hunk 12]
```diff
@@ -175,39 +182,41 @@ def request_worker_data(data_id):
...
+
+    received_data = 0
+    while received_data < len(data_ids):
+        # Blocking get
+        complex_data = pull_data(mpi_state.comm)
```
Comment (YarShev) [inline unidist/core/backends/mpi/core/controller/common.py:215]:
```suggestion
        complex_data = pull_data(mpi_state.comm, MPI.ANY_SOURCE)
```

File: unidist/core/backends/mpi/core/controller/common.py [hunk 13]
```diff
@@ -175,39 +182,41 @@ def request_worker_data(data_id):
...
+        data_id = complex_data["id"]
+        if data_id in data_ids:
```
Comment (YarShev) [inline unidist/core/backends/mpi/core/controller/common.py:217]: Can a worker send a data_id that is not in the list?

File: unidist/core/backends/mpi/core/controller/common.py [hunk 14]
```diff
@@ -175,39 +182,41 @@ def request_worker_data(data_id):
...
+            received_data += 1
+            data_id = [d_id for d_id in data_ids if d_id == data_id][0]
```
Comment (YarShev) [inline unidist/core/backends/mpi/core/controller/common.py:219]: We should probably have a dict here.

File: unidist/core/backends/mpi/core/controller/common.py [hunk 15]
```diff
@@ -175,39 +182,41 @@ def request_worker_data(data_id):
...
+        # completion of the communication, which is necessary for the pipeline to continue.
+        h_list = communication.isend_simple_operation(
+            mpi_state.comm,
+            operation_type,
+            operation_data,
+            owner_rank,
+        )
+        async_operations.extend(h_list)
+
+    received_data = 0
```
Comment (YarShev) [inline unidist/core/backends/mpi/core/controller/common.py:212]:
```suggestion
    data_count = 0
```

File: unidist/core/backends/mpi/core/controller/common.py [hunk 16]
```diff
@@ -124,7 +125,11 @@ def pull_data(comm, owner_rank):
     object
         Received data object from another MPI process.
     """
-    info_package = communication.mpi_recv_object(comm, owner_rank)
+    if owner_rank is None:
```
Comment (YarShev) [inline unidist/core/backends/mpi/core/controller/common.py:128]:
```suggestion
    if owner_rank == MPI.ANY_SOURCE:
```

File: unidist/core/backends/mpi/core/communication.py [hunk 17]
```diff
@@ -374,6 +374,33 @@ def mpi_recv_operation(comm):
...
```
[the remark this answers is not in the document]
Comment (Retribution98) [reply unidist/core/backends/mpi/core/communication.py:377]: done

File: unidist/core/backends/mpi/core/communication.py [hunk 18]
```diff
@@ -374,6 +374,33 @@ def mpi_recv_operation(comm):
...
+    """
+    Receive an object of a standard Python data type from any source.
+
+    Parameters
+    ----------
+    comm : object
```
[the remark this answers is not in the document]
Comment (Retribution98) [reply unidist/core/backends/mpi/core/communication.py:383]: done

File: unidist/core/backends/mpi/core/communication.py [hunk 19]
```diff
@@ -374,6 +374,33 @@ def mpi_recv_operation(comm):
...
+        MPI communicator object.
```
[the remark this answers is not in the document]
Comment (Retribution98) [reply unidist/core/backends/mpi/core/communication.py:384]: done

File: unidist/core/backends/mpi/core/communication.py [hunk 20]
```diff
@@ -374,6 +374,33 @@ def mpi_recv_operation(comm):
...
+
+    Returns
+    -------
+    object
+        Received data object from another MPI process.
+    int
+        Process rank from which the object was obtained.
```
[the remark this answers is not in the document]
Comment (Retribution98) [reply unidist/core/backends/mpi/core/communication.py:391]: done

File: unidist/core/backends/mpi/core/controller/api.py (hunk 4 above)
Comment (Retribution98) [reply unidist/core/backends/mpi/core/controller/api.py:421]: done

File: unidist/core/backends/mpi/core/controller/api.py (hunk 3 above)
Comment (Retribution98) [reply unidist/core/backends/mpi/core/controller/api.py:411]: done

File: unidist/core/backends/mpi/core/communication.py [hunk 21]
```diff
@@ -374,6 +374,33 @@ def mpi_recv_operation(comm):
     return op_type, status.Get_source()
 
 
+def mpi_iprobe_object(comm):
+    """
+    Receive an object of a standard Python data type from any source.
```
[the remark this answers is not in the document]
Comment (Retribution98) [reply unidist/core/backends/mpi/core/communication.py:379]: done

File: unidist/core/backends/mpi/core/controller/api.py (hunk 5 above)
Comment (Retribution98) [reply unidist/core/backends/mpi/core/controller/api.py:423]: done

File: unidist/core/backends/mpi/core/controller/api.py (hunk 6 above)
Comment (Retribution98) [reply unidist/core/backends/mpi/core/controller/api.py:432]: done

File: unidist/core/backends/mpi/core/controller/api.py (hunk 7 above)
Comment (Retribution98) [reply unidist/core/backends/mpi/core/controller/api.py:438]: done

File: unidist/core/backends/mpi/core/controller/common.py (hunk 16 above)
Comment (Retribution98) [reply unidist/core/backends/mpi/core/controller/common.py:128]: not actual

File: unidist/core/backends/mpi/core/controller/common.py (hunk 9 above)
Comment (Retribution98) [reply unidist/core/backends/mpi/core/controller/common.py:169]: done

File: unidist/core/backends/mpi/core/controller/common.py (hunk 10 above)
Comment (Retribution98) [reply unidist/core/backends/mpi/core/controller/common.py:203]: done

File: unidist/core/backends/mpi/core/controller/common.py (hunk 11 above)
Comment (Retribution98) [reply unidist/core/backends/mpi/core/controller/common.py:210]: No, because some data may be received before the sending operation is processed by another process.

File: unidist/core/backends/mpi/core/controller/common.py (hunk 15 above)
Comment (Retribution98) [reply unidist/core/backends/mpi/core/controller/common.py:212]: done

File: unidist/core/backends/mpi/core/controller/common.py (hunk 12 above)
Comment (Retribution98) [reply unidist/core/backends/mpi/core/controller/common.py:215]: not actual

File: unidist/core/backends/mpi/core/controller/common.py (hunk 13 above)
Comment (Retribution98) [reply unidist/core/backends/mpi/core/controller/common.py:217]: No, in this case I'll change it to throw an exception.

File: unidist/core/backends/mpi/core/controller/common.py (hunk 14 above)
Comment (Retribution98) [reply unidist/core/backends/mpi/core/controller/common.py:219]: We don't need this line at all.

File: unidist/core/backends/mpi/core/shared_object_store.py (hunk 2 above)
Comment (Retribution98) [reply unidist/core/backends/mpi/core/shared_object_store.py:434]: I am disagree, but it is ok.

**Review by YarShev — APPROVED** [review]
